### PR TITLE
Add VPC, instance profile, volume support to EC2LatentWorker

### PR DESCRIPTION
**Pull request by ryansydnor**

### CODE REVIEW

File: master/buildbot/test/unit/test_worker_ec2.py [hunk 1]
```diff
@@ -22,8 +24,6 @@
     boto = None
     ec2 = None
 
-from twisted.trial import unittest
```
Comment (sa2ajj) [inline master/buildbot/test/unit/test_worker_ec2.py]: why did you move the import?  lexicographically, `mock`, `moto`, `boto` and `ec2` come before `twisted`...

File: master/buildbot/worker/ec2.py [hunk 2]
```diff
@@ -71,7 +73,7 @@ def __init__(self, name, password, instance_type, ami=None,
                 "Use of default value of 'keypair_name' of EC2LatentWorker "
                 "constructor is deprecated. Please explicitly specify value")
             keypair_name = 'latent_buildbot_slave'
-        if security_name is None:
+        if security_name is None and not subnet_id:
             reportDeprecatedWorkerNameUsage(
```
Comment (sa2ajj) [inline master/buildbot/worker/ec2.py]: wow, I wonder how this is related to transition to worker... I think it should be just `DeprecationWarning` usage.  @rutsky?

File: master/docs/manual/cfg-workers-ec2.rst [hunk 3]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
                                tags={'SomeTag': 'foo'})
     ]
 
+If the worker needs access to additional AWS resources, you can also enable your workers to access them via an EC2 instance profile.
+To use this capability, you must first create an instance profile separately in AWS.  Then specify its name on EC2LatentWorker via instance_profile_name.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: Please put only one sentence per line.

File: master/docs/manual/cfg-workers-ec2.rst [hunk 4]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
+
 The :class:`~buildbot.worker.ec2.EC2LatentWorker` supports all other configuration from the standard :class:`Worker`.
 The ``missing_timeout`` and ``notify_on_missing`` specify how long to wait for an EC2 instance to attach before considering the attempt to have failed, and email addresses to alert, respectively.
 ``missing_timeout`` defaults to 20 minutes.
 
+
+Volumes
+--------------
+
+If you want to attach existing volumes to an ec2 latent worker, use the volumes attribute.  This mechanism can be valuable if you want to maintain state on a conceptual worker across multiple start/terminate sequences.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

File: master/docs/manual/cfg-workers-ec2.rst [hunk 5]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
 ``volumes`` expects a list of (volume_id, mount_point) tuples to attempt attaching when your instance has been created.
 
+If you want to attach new ephemeral volumes, use the the block_device_map attribute.  This follows the BlockDeviceMap
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

File: master/docs/manual/cfg-workers-ec2.rst [hunk 6]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
+configuration of boto almost exactly, essentially acting as a passthrough.  The only distinction is that the volumes default
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

File: master/docs/manual/cfg-workers-ec2.rst [hunk 7]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
+to deleting on termination to avoid leaking volume resources when workers are terminated.  See boto documentation for further details.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

File: master/docs/manual/cfg-workers-ec2.rst [hunk 8]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
+                                }
+                               }
+                               )
+    ]
+
+
+VPC Support
+--------------
+
+If you are managing workers within a VPC, your worker configuration must be modified from above.  You must specify the id of the subnet where you want your worker placed.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

File: master/docs/manual/cfg-workers-ec2.rst [hunk 9]
```diff
@@ -220,12 +220,74 @@ To get Buildbot to tag the latent worker specify the tag keys and values in your
...
+You must also specify security groups created within your VPC as opposed to classic EC2 security groups.  This can be done by passing the ids of the vpc security groups.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-ec2.rst]: ditto

**Comment by sa2ajj** [comment]
LGTM, except for one-sentence-per-line issues in the docs.

File: master/buildbot/worker/ec2.py (hunk 2 above)
Comment (rutsky) [inline master/buildbot/worker/ec2.py]: @sa2ajj during transition to worker terminology default `security_name` was changed from `'latent_buildbot_slave'` to "unspecified".
Helper function is used for easier managing of worker-related deprecations --- I hope we will remove them all at once at some point, plus it's not just `DeprecationWarning`, it's `DeprecatedWorkerNameWarning`.

@ryansydnor what would you do if `EC2LatentWorker.__init__` would have `security_name` keyword argument defaulted to `'latent_buildbot_slave'`?

```
class EC2LatentWorker(AbstractLatentWorker):
    def __init__(self, ..., security_name='latent_buildbot_slave', ...):
        ...
```

You would make it defaulted to `None` and check which one of `security_name` or `subnet_id` is specified?
Am I correct that `security_name` and `subnet_id` are exclusive options and one and only one option can be used?

File: master/buildbot/test/unit/test_worker_ec2.py (hunk 1 above)
Comment (ryansydnor) [inline master/buildbot/test/unit/test_worker_ec2.py]: pylint suggested it:

`C: 26, 0: external import "from twisted.trial import unittest" comes before "import boto" (wrong-import-order)`

File: master/buildbot/worker/ec2.py (hunk 2 above)
Comment (ryansydnor) [inline master/buildbot/worker/ec2.py]: 1. I'll leave DeprecatedWorkerNameUsage as is.
2. I don't think it makes sense to have a default value for security_name when it is an option only used outside of a VPC, something Amazon is deprecating and moving away from.
3. Yes, you are correct. `security_name` and `subnet_id` options are mutually exclusive. (as per this snippet)

```
if security_name and subnet_id:
            raise ValueError(
                'security_name (EC2 classic security groups) is not supported '
                'in a VPC.  Use security_group_ids instead.')
```

They indicate whether or not we will be using classic AWS EC2 or AWS VPC. Some more details are [here](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-vpc.html). The most relevant section of that, in my opinion:

> If you created your account after 2013-12-04, it supports EC2-VPC only

My preference is to leave the change as is, if you have any thoughts I'm open to hearing them!

File: master/docs/manual/cfg-workers-ec2.rst (hunk 3 above)
Comment (ryansydnor) [inline master/docs/manual/cfg-workers-ec2.rst]: done

File: master/docs/manual/cfg-workers-ec2.rst (hunk 8 above)
Comment (ryansydnor) [inline master/docs/manual/cfg-workers-ec2.rst]: done

File: master/docs/manual/cfg-workers-ec2.rst (hunk 7 above)
Comment (ryansydnor) [inline master/docs/manual/cfg-workers-ec2.rst]: done

File: master/docs/manual/cfg-workers-ec2.rst (hunk 9 above)
Comment (ryansydnor) [inline master/docs/manual/cfg-workers-ec2.rst]: done

**Comment by ryansydnor** [comment]
@sa2ajj - thanks for getting back to me so quickly! I've cleaned up the PR based on your comments. PTAL and let me know if there's anything else I can do to help push it through.

File: master/buildbot/worker/ec2.py (hunk 2 above)
Comment (rutsky) [inline master/buildbot/worker/ec2.py]: @ryansydnor thanks for the explanation. This part looks good for me.

Comment (sa2ajj) [inline master/buildbot/worker/ec2.py]: @rutsky thanks for the clarification: without that context it was unclear.

File: master/buildbot/test/unit/test_worker_ec2.py (hunk 1 above)
Comment (sa2ajj) [inline master/buildbot/test/unit/test_worker_ec2.py]: that's interesting. we need to look into that later then.

**Comment by sa2ajj** [comment]
:+1:

I'll give some time for another pair of prying eyes to have a look though.

**Comment by ryansydnor** [comment]
@tomprince @davebx @djmitche - any thoughts on this?

**Comment by sa2ajj** [comment]
@ryansydnor i think two days were long enough for anybody to find time to look at it :)

Thanks for the contribution!
